// @cc-plugin-codex/runtime/dist/errors.js
// Typed errors for the runtime job store. Each error carries enough structured context
// to drive future CLI messages and test assertions.
export class JobStoreError extends Error {
    name = 'JobStoreError';
    cause;
    constructor(message, cause) {
        super(message);
        this.cause = cause;
    }
}
export class JobNotFoundError extends JobStoreError {
    name = 'JobNotFoundError';
    jobId;
    constructor(jobId) {
        super(`Job not found: ${jobId}`);
        this.jobId = jobId;
    }
}
export class JobLockError extends JobStoreError {
    name = 'JobLockError';
    jobId;
    lockInfo;
    constructor(jobId, lockInfo) {
        super(`Job is locked: ${jobId}`);
        this.jobId = jobId;
        this.lockInfo = lockInfo;
    }
}
export class InvalidJobIdError extends JobStoreError {
    name = 'InvalidJobIdError';
    jobId;
    constructor(jobId) {
        super(`Invalid job ID: ${jobId}`);
        this.jobId = jobId;
    }
}
export class CorruptJobRecordError extends JobStoreError {
    name = 'CorruptJobRecordError';
    path;
    constructor(path, cause) {
        super(`Corrupt job record at ${path}`, cause);
        this.path = path;
    }
}
/**
 * Thrown by driver lifecycle methods that haven't been implemented yet in plan 0001.
 * The plan reference helps future readers grep for where the method lands.
 */
export class DriverNotImplementedError extends Error {
    name = 'DriverNotImplementedError';
    methodName;
    planReference;
    constructor(methodName, planReference) {
        const where = planReference ? ` (planned for ${planReference})` : '';
        super(`Driver method not implemented yet: ${methodName}${where}`);
        this.methodName = methodName;
        this.planReference = planReference;
    }
}
export class DriverError extends Error {
    name = 'DriverError';
    driverName;
    operation;
    exitCode;
    stdout;
    stderr;
    cause;
    constructor(message, ctx) {
        super(message);
        this.driverName = ctx.driverName;
        this.operation = ctx.operation;
        this.exitCode = ctx.exitCode;
        this.stdout = ctx.stdout;
        this.stderr = ctx.stderr;
        this.cause = ctx.cause;
    }
}
