// @codex-delegation/driver-agy-cli/dist/index.js
import { randomBytes } from 'node:crypto';
import { spawn } from 'node:child_process';
import { readFile, unlink, writeFile } from 'node:fs/promises';
import { basename, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { DriverError, ensureProviderSessionsDir } from '@codex-delegation/runtime';
import { probeAgyCliDriver } from './probe.js';
import { readAgyState, writeAgyState } from './state.js';
import { DRIVER_NAME } from './types.js';
export { probeAgyCliDriver } from './probe.js';
export { readAgyState } from './state.js';
export { DRIVER_NAME, DRIVER_VERSION } from './types.js';
const RUNNER_PATH = fileURLToPath(new URL('./runner.js', import.meta.url));
const TERMINAL = new Set(['completed', 'failed', 'stopped']);
function delay(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
}
function processExists(pid) {
    if (!pid || pid <= 0)
        return false;
    try {
        process.kill(pid, 0);
        return true;
    }
    catch (error) {
        return error.code === 'EPERM';
    }
}
function safeName(value) {
    return value.replace(/[^a-zA-Z0-9._:-]+/g, '-').slice(0, 80) || 'agy';
}
function modeFor(opts) {
    const permissionMode = opts.permissionMode === 'acceptEdits'
        ? 'accept-edits'
        : opts.permissionMode === 'plan'
            ? 'plan'
            : undefined;
    if (opts.permissionMode !== undefined &&
        !['default', 'acceptEdits', 'plan', 'bypassPermissions'].includes(opts.permissionMode)) {
        throw new DriverError(`agy does not support permission mode "${opts.permissionMode}"`, {
            driverName: DRIVER_NAME,
            operation: 'startSession',
        });
    }
    const requested = [opts.mode, permissionMode, opts.allowEdit ? 'accept-edits' : undefined].filter((value) => value !== undefined);
    if (new Set(requested).size > 1) {
        throw new DriverError('Conflicting Antigravity edit modes were requested', {
            driverName: DRIVER_NAME,
            operation: 'startSession',
        });
    }
    return requested[0];
}
function signalIfRunning(pid, signal) {
    if (!pid || !processExists(pid))
        return;
    try {
        process.kill(pid, signal);
    }
    catch (error) {
        if (error.code !== 'ESRCH')
            throw error;
    }
}
export function buildAgyArgs(opts) {
    const args = [];
    if (opts.model)
        args.push('--model', opts.model);
    if (opts.agent)
        args.push('--agent', opts.agent);
    const workspaceDirs = Array.from(new Set([opts.cwd, ...(opts.addDirs ?? [])].filter((dir) => typeof dir === 'string' && dir.length > 0)));
    for (const dir of workspaceDirs)
        args.push('--add-dir', dir);
    const mode = modeFor(opts);
    if (mode)
        args.push('--mode', mode);
    if (opts.sandbox || opts.safeMode)
        args.push('--sandbox');
    if (opts.dangerouslySkipPermissions || opts.permissionMode === 'bypassPermissions') {
        args.push('--dangerously-skip-permissions');
    }
    if (opts.printTimeout)
        args.push('--print-timeout', opts.printTimeout);
    if (opts.project)
        args.push('--project', opts.project);
    if (opts.newProject)
        args.push('--new-project');
    if (opts.logFile)
        args.push('--log-file', opts.logFile);
    args.push('--print', opts.prompt);
    return args;
}
async function waitForState(path, timeoutMs) {
    const deadline = Date.now() + timeoutMs;
    for (;;) {
        const state = await readAgyState(path);
        if (state !== null)
            return state;
        if (Date.now() >= deadline)
            return null;
        await delay(20);
    }
}
function handleStatePath(session) {
    if (!session.statePath) {
        throw new DriverError('agy session handle is missing statePath', {
            driverName: DRIVER_NAME,
            operation: 'session',
        });
    }
    return session.statePath;
}
function stateToStatus(state) {
    let value;
    switch (state.status) {
        case 'starting':
            value = 'starting';
            break;
        case 'running':
            value = 'working';
            break;
        case 'completed':
            value = 'completed';
            break;
        case 'failed':
            value = 'failed';
            break;
        case 'stopped':
            value = 'stopped';
            break;
    }
    return {
        value,
        shortId: state.shortId,
        sessionName: state.sessionName,
        cwd: state.cwd,
        pid: state.agyPid ?? state.runnerPid,
        startedAt: state.startedAt,
        updatedAt: state.updatedAt,
        raw: state,
    };
}
export async function readAgyOutput(path) {
    if (!path)
        return '';
    try {
        return await readFile(path, 'utf8');
    }
    catch (error) {
        if (error.code === 'ENOENT')
            return '';
        throw error;
    }
}
export class AgyCliDriver {
    defaults;
    constructor(options = {}) {
        this.defaults = options;
    }
    probe() {
        return probeAgyCliDriver(this.defaults);
    }
    async startSession(opts) {
        if (!opts.prompt.trim()) {
            throw new DriverError('startSession requires a non-empty prompt', {
                driverName: DRIVER_NAME,
                operation: 'startSession',
            });
        }
        const cwd = opts.cwd || this.defaults.cwd;
        if (!cwd) {
            throw new DriverError('startSession requires cwd', {
                driverName: DRIVER_NAME,
                operation: 'startSession',
            });
        }
        const shortId = randomBytes(4).toString('hex');
        const prefix = opts.name ? safeName(opts.name) : `codex:agy:${safeName(basename(cwd))}`;
        const sessionName = `${prefix}-${shortId}`;
        const sessionsDir = await ensureProviderSessionsDir('agy');
        const base = join(sessionsDir, shortId);
        const requestPath = `${base}.request.json`;
        const statePath = `${base}.state.json`;
        const resultPath = `${base}.stdout.txt`;
        const errorPath = `${base}.stderr.txt`;
        const startedAt = new Date().toISOString();
        const state = {
            schemaVersion: 1,
            driverName: DRIVER_NAME,
            shortId,
            sessionName,
            cwd,
            status: 'starting',
            runnerPid: 0,
            startedAt,
            updatedAt: startedAt,
            resultPath,
            errorPath,
        };
        const request = {
            executable: this.defaults.executable ?? this.defaults.env?.['AGY_CLI_PATH'] ?? 'agy',
            args: buildAgyArgs(opts),
            cwd,
            state,
        };
        await Promise.all([
            writeFile(requestPath, JSON.stringify(request), { mode: 0o600 }),
            writeFile(resultPath, '', { mode: 0o600 }),
            writeFile(errorPath, '', { mode: 0o600 }),
        ]);
        let runner;
        try {
            runner = spawn(process.execPath, [RUNNER_PATH, requestPath, statePath], {
                cwd,
                env: this.defaults.env ?? process.env,
                detached: true,
                shell: false,
                stdio: 'ignore',
            });
            await new Promise((resolve, reject) => {
                runner.once('spawn', resolve);
                runner.once('error', reject);
            });
            runner.unref();
        }
        catch (error) {
            await unlink(requestPath).catch(() => undefined);
            throw new DriverError('Failed to start the agy supervisor', {
                driverName: DRIVER_NAME,
                operation: 'startSession',
                cause: error,
            });
        }
        const observed = await waitForState(statePath, this.defaults.timeoutMs ?? 3000);
        if (observed?.status === 'failed') {
            const stderr = await readAgyOutput(errorPath);
            throw new DriverError(observed.error ?? 'agy failed to start', {
                driverName: DRIVER_NAME,
                operation: 'startSession',
                exitCode: observed.exitCode ?? undefined,
                stderr,
            });
        }
        if (observed === null) {
            if (runner.pid && processExists(runner.pid))
                runner.kill('SIGTERM');
            throw new DriverError('agy supervisor did not publish state before the startup timeout', {
                driverName: DRIVER_NAME,
                operation: 'startSession',
            });
        }
        return {
            driverName: DRIVER_NAME,
            shortId,
            sessionName,
            cwd,
            startedAt,
            pid: observed.agyPid ?? runner.pid,
            statePath,
            resultPath,
            errorPath,
        };
    }
    async *watch(target, opts = {}) {
        const session = 'session' in target ? target.session : target;
        let previous = '';
        for (;;) {
            if (opts.signal?.aborted)
                return;
            const status = await this.status(session);
            if (status.value !== previous) {
                yield { type: 'session.status', status, at: new Date().toISOString() };
                previous = status.value;
            }
            if (['completed', 'failed', 'stopped', 'orphaned'].includes(status.value))
                return;
            await delay(opts.intervalMs ?? 500);
        }
    }
    async status(session) {
        const state = await readAgyState(handleStatePath(session));
        if (state === null) {
            return {
                value: processExists(session.pid) ? 'starting' : 'orphaned',
                shortId: session.shortId,
                sessionName: session.sessionName,
                cwd: session.cwd,
                pid: session.pid,
                startedAt: session.startedAt,
            };
        }
        if (!TERMINAL.has(state.status) && !processExists(state.runnerPid)) {
            return { ...stateToStatus(state), value: 'orphaned' };
        }
        return stateToStatus(state);
    }
    async stop(session) {
        const statePath = handleStatePath(session);
        const state = await readAgyState(statePath);
        if (state === null || TERMINAL.has(state.status))
            return;
        const targetPid = state.runnerPid > 0 ? state.runnerPid : session.pid;
        if (targetPid && processExists(targetPid)) {
            try {
                process.kill(targetPid, 'SIGTERM');
            }
            catch (error) {
                if (error.code !== 'ESRCH')
                    throw error;
            }
        }
        const deadline = Date.now() + 3000;
        while (Date.now() < deadline) {
            const current = await readAgyState(statePath);
            if (current === null || current.status === 'stopped' || TERMINAL.has(current.status))
                return;
            await delay(50);
        }
        signalIfRunning(state.agyPid, 'SIGKILL');
        signalIfRunning(state.runnerPid, 'SIGKILL');
        const now = new Date().toISOString();
        await writeAgyState(statePath, {
            ...state,
            status: 'stopped',
            updatedAt: now,
            endedAt: now,
            signal: 'SIGKILL',
        });
    }
    send(_session, _input, _opts) {
        return Promise.reject(new DriverError('agy CLI follow-up is unavailable because print mode does not return a stable conversation ID; global --continue is intentionally not used.', { driverName: DRIVER_NAME, operation: 'send' }));
    }
    async dispose() { }
}
