// @codex-delegation/driver-agy-cli/dist/probe.js
import { spawn } from 'node:child_process';
import { DRIVER_NAME, DRIVER_VERSION } from './types.js';
function runAgy(args, options) {
    return new Promise((resolve) => {
        const child = spawn(options.executable ?? options.env?.['AGY_CLI_PATH'] ?? 'agy', args, {
            cwd: options.cwd,
            env: options.env ?? process.env,
            shell: false,
        });
        let stdout = '';
        let stderr = '';
        let error;
        let timedOut = false;
        const timer = setTimeout(() => {
            timedOut = true;
            child.kill('SIGKILL');
        }, options.timeoutMs ?? 5000);
        child.stdout.on('data', (chunk) => (stdout += chunk.toString()));
        child.stderr.on('data', (chunk) => (stderr += chunk.toString()));
        child.once('error', (value) => (error = value));
        child.once('close', (code) => {
            clearTimeout(timer);
            resolve({ code, stdout, stderr, error, timedOut });
        });
    });
}
function aggregate(probes) {
    if (probes.some((probe) => probe.status === 'fail'))
        return 'fail';
    if (probes.some((probe) => probe.status === 'warn'))
        return 'warn';
    return 'ok';
}
export async function probeAgyCliDriver(options = {}) {
    const versionResult = await runAgy(['--version'], options);
    const version = versionResult.stdout.trim() || versionResult.stderr.trim();
    const binaryOk = !versionResult.error && !versionResult.timedOut && versionResult.code === 0;
    const binaryProbe = binaryOk
        ? { name: 'agy-binary', status: 'ok', detail: version || 'agy is executable' }
        : {
            name: 'agy-binary',
            status: 'fail',
            detail: versionResult.timedOut
                ? 'agy --version timed out'
                : `Cannot run agy: ${versionResult.error?.code || versionResult.stderr.trim() || 'unknown error'}`,
        };
    const helpResult = binaryOk
        ? await runAgy(['--help'], options)
        : { code: null, stdout: '', stderr: '', timedOut: false };
    const help = `${helpResult.stdout}\n${helpResult.stderr}`;
    const printOk = helpResult.code === 0 && /--print(?:\s|string|,)/.test(help);
    const printProbe = printOk
        ? { name: 'agy-print-mode', status: 'ok', detail: 'agy --print is available' }
        : {
            name: 'agy-print-mode',
            status: 'fail',
            detail: binaryOk ? 'agy --help does not advertise --print' : 'agy binary unavailable',
        };
    const authProbe = {
        name: 'agy-auth',
        status: 'warn',
        detail: 'Authentication is checked on first delegation to avoid a token-spending probe.',
    };
    const probes = [binaryProbe, printProbe, authProbe];
    return {
        driverName: DRIVER_NAME,
        driverVersion: DRIVER_VERSION,
        cliVersion: binaryOk ? version : null,
        execution: 'supervised-process',
        features: {
            start: printOk,
            status: true,
            stop: true,
            followup: false,
            logs: true,
        },
        backgroundSessions: printOk,
        agentsJson: false,
        logsCommand: true,
        transcriptPath: false,
        attach: false,
        structuredStream: 'output',
        toolEvents: 'none',
        permissions: 'cli-policy',
        health: { status: aggregate(probes), probes },
    };
}
