// @cc-plugin-codex/runtime/dist/paths.js
// Filesystem path helpers for the cc-plugin-codex companion.
//
// Default state lives under `~/.codex/cc-plugin-codex/`. Tests MUST override via the
// CC_PLUGIN_CODEX_HOME env var to an isolated temp directory so they do not touch real state.
//
// Path helpers do NOT create directories on their own. The only function that touches the
// filesystem is `ensureCompanionDirs()`, called explicitly by job-store entry points.
import { mkdir } from 'node:fs/promises';
import { homedir } from 'node:os';
import { join } from 'node:path';
const ENV_VAR = 'CC_PLUGIN_CODEX_HOME';
function defaultHome() {
    return join(homedir(), '.codex', 'cc-plugin-codex');
}
export function getCompanionHome() {
    const override = process.env[ENV_VAR];
    return override && override.length > 0 ? override : defaultHome();
}
export function getJobsDir() {
    return join(getCompanionHome(), 'jobs');
}
export function getLogsDir() {
    return join(getCompanionHome(), 'logs');
}
export function getDoctorPath() {
    return join(getCompanionHome(), 'doctor.json');
}
export function getJobRecordPath(jobId) {
    return join(getJobsDir(), `${jobId}.json`);
}
export function getJobEventsPath(jobId) {
    return join(getJobsDir(), `${jobId}.events.jsonl`);
}
export function getJobResultPath(jobId) {
    return join(getJobsDir(), `${jobId}.result.md`);
}
export function getJobLockPath(jobId) {
    return join(getJobsDir(), `${jobId}.lock`);
}
export async function ensureCompanionDirs() {
    await mkdir(getJobsDir(), { recursive: true });
    await mkdir(getLogsDir(), { recursive: true });
}
