// @codex-delegation/driver-qwen-code/dist/index.js
import { randomBytes } from 'node:crypto';
import { spawn } from 'node:child_process';
import { mkdir, readFile, writeFile } from 'node:fs/promises';
import { basename, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { DriverError, ensureProviderSessionsDir } from '@codex-delegation/runtime';
import { probeQwenCodeDriver } from './probe.js';
import { acquireQwenSendLock, readQwenState, writeQwenState } from './state.js';
import { finalQwenAssistantMessage, readQwenTranscriptEvents } from './transcript.js';
import { DRIVER_NAME } from './types.js';
export { probeQwenCodeDriver } from './probe.js';
export { readQwenState, writeQwenState } from './state.js';
export * from './transcript.js';
export { DRIVER_NAME, DRIVER_VERSION } from './types.js';
const RUNNER_PATH = fileURLToPath(new URL('./runner.js', import.meta.url));
const TERMINAL = new Set(['idle', 'failed', 'stopped']);
function delay(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
}
function exists(pid) {
    if (!pid || pid <= 0)
        return false;
    try {
        process.kill(pid, 0);
        return true;
    }
    catch (error) {
        return error.code === 'EPERM';
    }
}
function safeName(value) {
    return value.replace(/[^a-zA-Z0-9._:-]+/g, '-').slice(0, 80) || 'qwen';
}
function approvalMode(opts) {
    if (opts.dangerouslySkipPermissions || opts.permissionMode === 'bypassPermissions')
        return 'yolo';
    if (opts.permissionMode === 'plan')
        return 'plan';
    if (opts.permissionMode === 'acceptEdits' || opts.allowEdit)
        return 'auto-edit';
    if (opts.permissionMode === undefined || opts.permissionMode === 'default')
        return undefined;
    if (opts.permissionMode === 'auto')
        return 'auto';
    throw new DriverError(`qwen does not support permission mode "${opts.permissionMode}"`, {
        driverName: DRIVER_NAME,
        operation: 'startSession',
    });
}
export function buildQwenResumeArgs(opts) {
    const args = ['--output-format', 'stream-json'];
    if (opts.model)
        args.push('--model', opts.model);
    if (opts.systemPrompt)
        args.push('--system-prompt', opts.systemPrompt);
    if (opts.appendSystemPrompt)
        args.push('--append-system-prompt', opts.appendSystemPrompt);
    if (opts.sandbox)
        args.push('--sandbox');
    if (opts.safeMode)
        args.push('--safe-mode');
    const mode = approvalMode(opts);
    if (mode)
        args.push('--approval-mode', mode);
    for (const dir of opts.addDirs ?? [])
        args.push('--include-directories', dir);
    for (const tool of opts.allowedTools ?? [])
        args.push('--allowed-tools', tool);
    for (const tool of opts.disallowedTools ?? [])
        args.push('--exclude-tools', tool);
    return args;
}
export function buildQwenArgs(opts) {
    const args = buildQwenResumeArgs(opts);
    if (opts.conversationId)
        args.push('--resume', opts.conversationId);
    args.push(opts.prompt);
    return args;
}
async function waitForState(path, predicate, timeoutMs) {
    const deadline = Date.now() + timeoutMs;
    for (;;) {
        const state = await readQwenState(path);
        if (state && predicate(state))
            return state;
        if (Date.now() >= deadline)
            return state;
        await delay(20);
    }
}
function stateStatus(state) {
    const value = state.status === 'running'
        ? 'working'
        : state.status === 'starting'
            ? 'starting'
            : state.status;
    return {
        value,
        shortId: state.shortId,
        sessionId: state.sessionId,
        sessionName: state.sessionName,
        cwd: state.cwd,
        pid: state.childPid ?? state.runnerPid,
        startedAt: state.startedAt,
        updatedAt: state.updatedAt,
        transcriptPath: state.transcriptPath,
        raw: state,
    };
}
export async function readQwenOutput(path) {
    if (!path)
        return '';
    try {
        return await readFile(path, 'utf8');
    }
    catch (error) {
        if (error.code === 'ENOENT')
            return '';
        throw error;
    }
}
export class QwenCodeDriver {
    defaults;
    constructor(defaults = {}) {
        this.defaults = defaults;
    }
    probe() {
        return probeQwenCodeDriver(this.defaults);
    }
    async launch(statePath, state, args, turnIndex) {
        const requestPath = `${statePath}.request.${turnIndex}.json`;
        const request = {
            executable: this.defaults.executable ?? this.defaults.env?.['QWEN_CLI_PATH'] ?? 'qwen',
            args,
            cwd: state.cwd,
            env: { ...process.env, ...this.defaults.env },
            state: { ...state, status: 'starting', turnIndex, updatedAt: new Date().toISOString() },
            turnIndex,
        };
        await writeFile(requestPath, JSON.stringify(request), { mode: 0o600 });
        const runner = spawn(process.execPath, [RUNNER_PATH, requestPath, statePath], {
            cwd: state.cwd,
            env: process.env,
            detached: true,
            shell: false,
            stdio: 'ignore',
        });
        await new Promise((resolve, reject) => {
            runner.once('spawn', resolve);
            runner.once('error', reject);
        });
        runner.unref();
        return runner;
    }
    async startSession(opts) {
        if (!opts.prompt.trim()) {
            throw new DriverError('startSession requires a non-empty prompt', {
                driverName: DRIVER_NAME,
                operation: 'startSession',
            });
        }
        const cwd = opts.cwd || this.defaults.cwd;
        if (!cwd) {
            throw new DriverError('startSession requires cwd', {
                driverName: DRIVER_NAME,
                operation: 'startSession',
            });
        }
        const shortId = randomBytes(4).toString('hex');
        const sessionName = `${opts.name ? safeName(opts.name) : `codex:qwen:${safeName(basename(cwd))}`}-${shortId}`;
        const base = join(await ensureProviderSessionsDir('qwen'), shortId);
        const statePath = `${base}.state.json`;
        const resultPath = `${base}.transcript.jsonl`;
        const errorPath = `${base}.stderr.txt`;
        await Promise.all([
            mkdir(join(base, '..'), { recursive: true, mode: 0o700 }),
            writeFile(resultPath, '', { mode: 0o600 }),
            writeFile(errorPath, '', { mode: 0o600 }),
        ]);
        const startedAt = new Date().toISOString();
        const state = {
            schemaVersion: 1,
            driverName: DRIVER_NAME,
            shortId,
            sessionName,
            cwd,
            status: 'starting',
            runnerPid: 0,
            startedAt,
            updatedAt: startedAt,
            resultPath,
            errorPath,
            transcriptPath: resultPath,
            resumeArgs: buildQwenResumeArgs(opts),
            turnIndex: 0,
        };
        await writeQwenState(statePath, state);
        const runner = await this.launch(statePath, state, buildQwenArgs(opts), 0);
        const observed = await waitForState(statePath, (candidate) => candidate.runnerPid > 0, this.defaults.timeoutMs ?? 3000);
        if (!observed?.runnerPid) {
            if (exists(runner.pid))
                runner.kill('SIGTERM');
            throw new DriverError('qwen supervisor did not publish state', {
                driverName: DRIVER_NAME,
                operation: 'startSession',
            });
        }
        return {
            driverName: DRIVER_NAME,
            shortId,
            ...(observed.sessionId ? { sessionId: observed.sessionId } : {}),
            sessionName,
            cwd,
            startedAt,
            pid: observed.runnerPid,
            statePath,
            resultPath,
            errorPath,
        };
    }
    async status(session) {
        if (!session.statePath) {
            throw new DriverError('qwen session handle is missing statePath', {
                driverName: DRIVER_NAME,
                operation: 'status',
            });
        }
        const state = await readQwenState(session.statePath);
        if (!state)
            return { value: exists(session.pid) ? 'starting' : 'orphaned' };
        if (!TERMINAL.has(state.status) && !exists(state.runnerPid)) {
            return { ...stateStatus(state), value: 'orphaned' };
        }
        return stateStatus(state);
    }
    async *watch(target, opts = {}) {
        const session = 'session' in target ? target.session : target;
        let previous = '';
        for (;;) {
            if (opts.signal?.aborted)
                return;
            const status = await this.status(session);
            if (status.value !== previous) {
                yield { type: 'session.status', status, at: new Date().toISOString() };
                previous = status.value;
            }
            if (['idle', 'failed', 'stopped', 'orphaned'].includes(status.value))
                return;
            await delay(opts.intervalMs ?? 250);
        }
    }
    async stop(session) {
        if (!session.statePath)
            return;
        const state = await readQwenState(session.statePath);
        if (!state || TERMINAL.has(state.status))
            return;
        if (exists(state.runnerPid))
            process.kill(state.runnerPid, 'SIGTERM');
        const deadline = Date.now() + (this.defaults.stopTimeoutMs ?? 3000);
        while (Date.now() < deadline) {
            const current = await readQwenState(session.statePath);
            if (!current || TERMINAL.has(current.status))
                return;
            await delay(40);
        }
        const childPid = state.childPid;
        if (exists(childPid))
            process.kill(childPid, 'SIGKILL');
        if (exists(state.runnerPid))
            process.kill(state.runnerPid, 'SIGKILL');
        const endedAt = new Date().toISOString();
        await writeQwenState(session.statePath, {
            ...state,
            status: 'stopped',
            signal: 'SIGKILL',
            updatedAt: endedAt,
            endedAt,
        });
    }
    async send(session, input, opts = {}) {
        if (input.type !== 'text' || !input.text.trim() || !session.statePath) {
            throw new DriverError('qwen follow-up requires text and persisted state', {
                driverName: DRIVER_NAME,
                operation: 'send',
            });
        }
        const release = await acquireQwenSendLock(session.statePath).catch((error) => {
            throw new DriverError(error.message, {
                driverName: DRIVER_NAME,
                operation: 'send',
                cause: error,
            });
        });
        const startedAt = new Date().toISOString();
        try {
            const previous = await readQwenState(session.statePath);
            if (!previous || previous.status !== 'idle') {
                throw new DriverError(`qwen job is ${previous?.status ?? 'missing'}`, {
                    driverName: DRIVER_NAME,
                    operation: 'send',
                });
            }
            if (!previous.sessionId) {
                throw new DriverError('qwen session identity was not captured', {
                    driverName: DRIVER_NAME,
                    operation: 'send',
                });
            }
            const target = previous.turnIndex + 1;
            const args = [...previous.resumeArgs, '--resume', previous.sessionId, input.text];
            await this.launch(session.statePath, { ...previous, endedAt: undefined }, args, target);
            const deadline = Date.now() + (opts.timeoutMs ?? 600_000);
            for (;;) {
                if (opts.signal?.aborted) {
                    throw new DriverError('qwen follow-up wait aborted; turn remains active', {
                        driverName: DRIVER_NAME,
                        operation: 'send',
                    });
                }
                const current = await readQwenState(session.statePath);
                if (!current)
                    throw new Error('qwen state disappeared');
                if (current.status === 'failed' || current.status === 'stopped') {
                    throw new DriverError(current.error ?? `qwen follow-up ${current.status}`, {
                        driverName: DRIVER_NAME,
                        operation: 'send',
                        exitCode: current.exitCode ?? undefined,
                        stderr: await readQwenOutput(current.errorPath),
                    });
                }
                if (current.status === 'idle' && current.completedTurnIndex === target) {
                    const transcript = await readQwenTranscriptEvents({
                        transcriptPath: current.transcriptPath,
                    });
                    return {
                        driverName: DRIVER_NAME,
                        session: { ...session, sessionId: current.sessionId },
                        startedAt,
                        endedAt: new Date().toISOString(),
                        status: 'completed',
                        finalMessage: finalQwenAssistantMessage(transcript.events),
                    };
                }
                if (Date.now() >= deadline) {
                    throw new DriverError('qwen follow-up timed out; turn remains active', {
                        driverName: DRIVER_NAME,
                        operation: 'send',
                    });
                }
                await delay(50);
            }
        }
        finally {
            await release();
        }
    }
    async dispose() { }
}
